fuel-types/src/canonical.rs

```diff
@@ -0,0 +1,688 @@
+        validate(u64::MAX);
+        validate(123u128);
+        validate(u128::MAX);
+        validate(Vec::<u8>::new());
+        validate(Vec::<u16>::new());
+        validate(Vec::<u32>::new());
+        validate(Vec::<u64>::new());
+        validate(Vec::<u128>::new());
+        validate(vec![1u8]);
+        validate(vec![1u16]);
+        validate(vec![1u32]);
+        validate(vec![1u64]);
+        validate(vec![1u128]);
+        validate(vec![1u8, 2u8]);
+        validate(vec![1u16, 2u16]);
+        validate(vec![1u32, 2u32]);
+        validate(vec![1u64, 2u64]);
+        validate(vec![1u128, 2u128]);
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+        struct TestStruct1 {
+            a: u8,
+            b: u16,
+        }
+
+        validate(TestStruct1 { a: 123, b: 456 });
+        assert_eq!(TestStruct1::SIZE_STATIC, 16);
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+        struct TestStruct2 {
+            a: u8,
+            v: Vec<u8>,
+            b: u16,
+            arr0: [u8; 0],
+            arr1: [u8; 2],
+            arr2: [u16; 3],
+            arr3: [u64; 4],
+        }
+
+        validate(TestStruct2 {
+            a: 123,
+            v: vec![1, 2, 3],
+            b: 456,
+            arr0: [],
+            arr1: [1, 2],
+            arr2: [1, 2, u16::MAX],
+            arr3: [0, 3, 1111, u64::MAX],
+        });
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+        #[repr(transparent)]
+        struct TestStruct3([u8; 64]);
+
+        validate(TestStruct3([1; 64]));
+        assert_eq!(TestStruct3::SIZE_STATIC, 64);
+
+        #[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+        #[canonical(prefix = 1u64)]
+        struct Prefixed1 {
+            a: [u8; 3],
+            b: Vec<u8>,
+        }
+        validate(Prefixed1 {
+            a: [1, 2, 3],
+            b: vec![4, 5, 6],
+        });
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+        #[repr(u8)]
+        enum TestEnum1 {
+            A,
+            B,
+        }
+
+        validate(TestEnum1::A);
+        validate(TestEnum1::B);
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+        enum TestEnum2 {
+            A(u8),
+            B([u8; 3]),
+            C(Vec<u8>),
+        }
+
+        validate_enum(TestEnum2::A(2));
+        validate_enum(TestEnum2::B([1, 2, 3]));
+        validate_enum(TestEnum2::C(vec![1, 2, 3]));
+
+        #[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
+        #[canonical(prefix = 2u64)]
+        struct Prefixed2(u16);
+        validate(Prefixed2(u16::MAX));
+
+        #[derive(Debug, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+        #[repr(u64)]
+        enum TestEnum3Discriminant {
+            P1 = 1,
+            P2 = 2,
+        }
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+        #[canonical(inner_discriminant = TestEnum3Discriminant)]
```

Comment:
should we also have a test for regular a non-inner discriminant, ie. like we use on Input?

```rust
#[canonical(discriminant = InputRepr)]
```

Reply:
Yes. Fortunately it was even easier than that; I simply removed it completely, since it was only used when manually deserializing anyway. 1e14e956